llvm/lib/CAS/ActionCaches.cpp

```diff
@@ -235,17 +235,30 @@ Error UnifiedOnDiskActionCache::putImpl(ArrayRef<uint8_t> Key,
 }
 
 Error UnifiedOnDiskActionCache::validate() const {
-  auto ValidateRef = [](FileOffset Offset, ArrayRef<char> Value) -> Error {
+  auto ValidateRef = [this](FileOffset Offset, ArrayRef<char> Value) -> Error {
     auto ID = ondisk::UnifiedOnDiskCache::getObjectIDFromValue(Value);
     auto formatError = [&](Twine Msg) {
       return createStringError(
           llvm::errc::illegal_byte_sequence,
           "bad record at 0x" +
-              utohexstr((unsigned)Offset.get(), /*LowerCase=*/true) + ": " +
-              Msg.str());
+              utohexstr((unsigned)Offset.get(), /*LowerCase=*/true) +
+              " ref=0x" + utohexstr(ID.getOpaqueData(), /*LowerCase=*/true) +
+              ": " + Msg.str());
     };
     if (ID.getOpaqueData() == 0)
       return formatError("zero is not a valid ref");
+    // Use containsObject first, because getDigest assumes a valid ObjectID.
+    if (!UniDB->getGraphDB().containsObject(ID))
```

Comment:
Maybe we can just make a version of `getDigest` that returns `Expected`?

Reply:
`friend` class sounds fine. `getObjectPresence` might be useful as well since it can do containsObject check in primary. Which is a useful feature as well.

I don't hold strong in any of the suggestions since finding a ref in upstream will like fail the last check. But if we can avoid all that and make validation faster, it is better.

Reply:
Friend doesn't work because of unnamed namespace.  I kept the current code but introduced `CheckUpstream` parameters on the API so we don't mix CAS instances.